replicator/index.js

```diff
@@ -0,0 +1,76 @@
+import {
+  GetObjectCommand,
+  HeadObjectCommand,
+  PutObjectCommand,
+} from '@aws-sdk/client-s3'
+
+/**
+ * @typedef {import('@aws-sdk/client-s3').S3Client} S3Client
+ */
+
+/**
+ * Replicate event target into destination bucket and write a side index for it.
+ *
+ * @param {object} props
+ * @param {import('./utils/parse-sqs-event').EventRecord} props.record
+ * @param {S3Client} props.destinationBucket
+ * @param {S3Client} props.originBucket
+ * @param {string} props.destinationBucketName
+ */
+ export const replicate = async ({
+  record,
+  destinationBucket,
+  destinationBucketName,
+  originBucket,
+}) => {
+  const key = record.key
+
+  // Verify if event file already exist in destination bucket
+  try {
+    await destinationBucket.send(
+      new HeadObjectCommand({
+        Bucket: destinationBucketName,
+        Key: key,
+      })
+    )
+  } catch {
+    // Not in destinationBucket, so read from origin bucket and write to destination bucket
+    const getCmd = new GetObjectCommand({
+      Bucket: record.bucketName,
+      Key: key,
+    })
+
+    // TODO: md5
```

Comment:
We should DO this TODO!

I think the aws sdk does md5 validation automatically, but this wasn't always the case, so we should consider doing our own validation as we do in the dotStorage workers.

We could require that all files are written to buckets with the `sha256` as a metadata property, so we can reuse it here regardless of key structure or which bucket it's written to.

Reply:
Yes, we definitely will do this before launch. Created an issue https://github.com/web3-storage/upload-api/issues/58